src/py/flwr/server/superlink/driver/serverappio_servicer_test.py

```diff
@@ -54,3 +86,332 @@ def test_raise_if_true() -> None:
+        ]
+    )  # type: ignore
+    def test_get_nodes_not_successful_if_not_running(
+        self, num_transitions: int
+    ) -> None:
+        """Test `GetNodes` not sucessful if RunStatus is pending."""
+        # Prepare
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+
+        self._transition_run_status(run_id, num_transitions)
+
+        # Execute & Assert
+        self._assert_get_nodes_not_allowed(run_id)
+
+    def test_successful_push_task_ins_if_running(self) -> None:
+        """Test `PushTaskIns` success."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+        task_ins = create_task_ins(
+            consumer_node_id=node_id, anonymous=False, run_id=run_id
+        )
+
+        # Transition status to running. PushTaskRes is only allowed in running status.
+        self._transition_run_status(run_id, 2)
+        request = PushTaskInsRequest(task_ins_list=[task_ins], run_id=run_id)
+
+        # Execute
+        response, call = self._push_task_ins.with_call(request=request)
+
+        # Assert
+        assert isinstance(response, PushTaskInsResponse)
+        assert grpc.StatusCode.OK == call.code()
+
+    def _assert_push_task_ins_not_allowed(self, task_ins: TaskIns, run_id: int) -> None:
+        """Assert `PushTaskIns` not allowed."""
+        run_status = self.state.get_run_status({run_id})[run_id]
+        request = PushTaskInsRequest(task_ins_list=[task_ins], run_id=run_id)
+
+        with self.assertRaises(grpc.RpcError) as e:
+            self._push_task_ins.with_call(request=request)
+        assert e.exception.code() == grpc.StatusCode.PERMISSION_DENIED
+        assert e.exception.details() == self.status_to_msg[run_status.status]
+
+    @parameterized.expand(
+        [
+            (0,),  # Test not successful if RunStatus is pending.
+            (1,),  # Test not successful if RunStatus is starting.
+            (3,),  # Test not successful if RunStatus is finished.
+        ]
+    )  # type: ignore
+    def test_push_task_ins_not_successful_if_not_running(
+        self, num_transitions: int
+    ) -> None:
+        """Test `PushTaskIns` not successful if RunStatus is not running."""
+        # Prepare
+        node_id = self.state.create_node(ping_interval=30)
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+        task_ins = create_task_ins(
+            consumer_node_id=node_id, anonymous=False, run_id=run_id
+        )
+
+        self._transition_run_status(run_id, num_transitions)
+
+        # Execute & Assert
+        self._assert_push_task_ins_not_allowed(task_ins, run_id)
+
+    def test_pull_task_res_successful_if_running(self) -> None:
+        """Test `PullTaskRes` success."""
+        # Prepare
+        run_id = self.state.create_run("", "", "", {}, ConfigsRecord())
+        task_ids = self.state.get_task_ids_from_run_id(run_id)
```

Comment:
This will be an empty set, no? Are the TaskIns inserted somewhere in the code?

Reply:
Second thought, this should be fine for this test. We've had tests for get_task_res in `linkstate_test.py`. I will resolve this.